extension_templates/forecasting.py

```diff
@@ -260,6 +283,55 @@ def _predict_quantiles(self, fh, X=None, alpha=0.5):
                 at quantile probability in second-level col index, for each row index.
         """
         # implement here
+        # IMPORTANT: avoid side effects to y, X, fh, alpha
+        #
+        # Note: unlike in predict_quantiles where alpha can be float or list of float
+        #   alpha in _predict_quantiles is guaranteed to be a list of float
+
+    # implement one of _predict_interval or _predict_quantiles (above), or delete both
+    #
+    # if implementing _predict_quantiles, delete _predict_interval
+    # if not implementing either, delete both methods
+    def _predict_interval(self, fh, X=None, coverage=None):
+        """Compute/return prediction quantiles for a forecast.
+
+        private _predict_interval containing the core logic,
+            called from predict_interval and possibly predict_quantiles
+
+        State required:
+            Requires state to be "fitted".
+
+        Accesses in self:
+            Fitted model attributes ending in "_"
+            self.cutoff
+
+        Parameters
+        ----------
+        fh : int, list, np.array or ForecastingHorizon
+            Forecasting horizon, default = y.index (in-sample forecast)
+        X : pd.DataFrame, optional (default=None)
+            Exogenous time series
+        coverage : list of float (guaranteed not None and floats in [0,1] interval)
+           nominal coverage(s) of predictive interval(s)
+
+        Returns
+        -------
+        pred_int : pd.DataFrame
+            Column has multi-index: first level is variable name from y in fit,
+                second level coverage fractions for which intervals were computed.
+                    in the same order as in input `coverage`.
+                Third level is string "lower" or "upper", for lower/upper interval end.
+            Row index is fh. Entries are forecasts of lower/upper interval end,
+                for var in col index, at nominal coverage in second col index,
+                lower/upper depending on third col index, for the row index.
+                Upper/lower interval end forecasts are equivalent to
+                quantile forecasts at alpha = 0.5 - c/2, 0.5 + c/2 for c in coverage.
+        """
+        # implement here
+        # IMPORTANT: avoid side effects to y, X, fh, coverage
```

Comment:
side effects should be solved in parent class, like `X = X.copy()`? before calling the private method of the child?

Reply:
this would be a general question about `predict` and all.

I'd say: no, copying around wastes memory - it's perfectly fine passing references if they are not mutated, so we should not enforce a copy.
We have a test which checks whether there are side effects to arguments, `test_methods_have_no_side_effects`